### PR TITLE
Implemented Magic Wand Tool

### LINKED ISSUE #193: Magic Wand tool
Selects all pixels that have the same color. It should have those options:
- Selection Modes: Normal, Add, Subtract, Intersect
- Global/Continous. Global looks through the whole image, Continuous looks through connected pixels (using flood fill or similar)
- Sampling: Layer/Image
- Tolerance: How close a color should be to the clicked color to be selected

### PR DESCRIPTION
**Pull request by CPKreuz**
Closes #193, #197 and #200

### CODE REVIEW

**Review by Equbuxu — DISMISSED** [review]
zoom tool actiondisplay doesn't work (pressing down alt doesn't change the hint)
make a 9999*9999 canvas -> click with the magic wand -> pixieditor hangs
please implement other things from the magic wand issue

also maybe don't implement them just yet because the selection is likely to be rewritten together with the move tool and I think flabbet was going to work on that soon

**Comment by CPKreuz** [comment]
> zoom tool actiondisplay doesn't work (pressing down alt doesn't change the hint)
This actually appears to be because windows is selecting the menu bar

> make a 9999*9999 canvas -> click with the magic wand -> pixieditor hangs
I guess that's just an general flood fill issue

**Review by flabbet — CHANGES_REQUESTED** [review]
One feature I would definitely add, is multi-layer support. Tolerance would be good too, but it's all right if it's not here, since we are working with pixel-art images.

File: PixiEditor/Models/ImageManipulation/BitmapUtils.cs [hunk 1]
```diff
@@ -40,16 +40,16 @@ public static WriteableBitmap BytesToWriteableBitmap(int currentBitmapWidth, int
...
             WriteableBitmap finalBitmap = BitmapFactory.New(width, height);
 
             using (finalBitmap.GetBitmapContext())
             {
-                for (int i = 0; i < layers.Length; i++)
+                for (int i = 0; i < layers.Count(); i++)
                 {
-                    float layerOpacity = structure == null ? layers[i].Opacity : LayerStructureUtils.GetFinalLayerOpacity(layers[i], structure);
-                    Layer layer = layers[i];
+                    float layerOpacity = structure == null ? layers.ElementAt(i).Opacity : LayerStructureUtils.GetFinalLayerOpacity(layers.ElementAt(i), structure);
```
Comment (flabbet) [inline PixiEditor/Models/ImageManipulation/BitmapUtils.cs:51]: `ElementAt(i)` can be called once and saved into variable, it's more efficient.

File: PixiEditor/Models/Tools/Tools/MagicWandTool.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            Document document = BitmapManager.ActiveDocument;
+            Layer layer;
+
+            if (documentScope == DocumentScope.SingleLayer)
+            {
+                layer = BitmapManager.ActiveLayer;
+            }
+            else
+            {
+                layer = new Layer("_CombinedLayers", BitmapUtils.CombineLayers(document.Width, document.Height, document.Layers, document.LayerStructure));
```
Comment (flabbet) [inline PixiEditor/Models/Tools/Tools/MagicWandTool.cs:48]: Implement caching, cache should be saved until tool is changed.

**Review by flabbet — CHANGES_REQUESTED** [review]

File: PixiEditor/Models/Tools/Tools/MagicWandTool.cs (hunk 2 above)
Comment (CPKreuz) [reply PixiEditor/Models/Tools/Tools/MagicWandTool.cs:48]: What if the document/canvas get's resized?

Comment (flabbet) [reply PixiEditor/Models/Tools/Tools/MagicWandTool.cs:48]: Well of course, you need to consider this kind of scenarios

Comment (flabbet) [reply PixiEditor/Models/Tools/Tools/MagicWandTool.cs:48]: In that case, cache should be refreshed

**Review by flabbet — APPROVED** [review]
flabbet certified

**Review by flabbet — CHANGES_REQUESTED** [review]
I was too fast, the moment you shutdown PixiEditor, it crashes
image

**Review by flabbet — CHANGES_REQUESTED** [review]
Fix tests :p

**Review by flabbet — APPROVED** [review]

**Pull request merged**